xonsh/ptk_shell/formatter.py

```diff
@@ -0,0 +1,68 @@
+import builtins
+import functools
+from collections import defaultdict
+from concurrent.futures import ThreadPoolExecutor
+
+from xonsh.prompt.base import PromptFormatter, DEFAULT_PROMPT
+
+
+class PTKPromptFormatter(PromptFormatter):
+    def __init__(self):
+        super().__init__()
+
+        # the attribute, .cache is cleared between calls. This caches results from callback alone by field name.
+        self.thread_results = {}
+        self.futures = defaultdict(dict)
+
+    def __call__(self, template=DEFAULT_PROMPT, fields=None, threaded=False):
+        """Formats a xonsh prompt template string."""
+
+        # set these attributes per call. one can enable/disable async-prompt inside a session.
+        self.template = template  # current template
+        self.threaded = threaded
+
+        # clear out old futures from the same prompt
+        for fut in self.futures[template]:
+            fut.cancel()
+        self.futures[template].clear()
+
+        # init pool only on async prompts
+        if threaded and (not hasattr(self, "thread_pool")):
+            self.thread_pool = ThreadPoolExecutor(
+                max_workers=builtins.__xonsh__.env["ASYNC_PROMPT_THREAD_WORKERS"]
+            )
+
+        prompt = super().__call__(template, fields)
+
+        if threaded:
+            return prompt, self.futures[template]
+        return prompt
+
+    def _format_prompt(self, template=DEFAULT_PROMPT):
+        tokens = self._get_tokens(template)
+        if self.threaded:
+            return tokens
+        return "".join(tokens)
+
+    def _no_cache_field_value(self, field, field_value, idx: int = None):
+        func = functools.partial(super()._no_cache_field_value, field, field_value, idx)
+
+        if self.threaded and callable(field_value):
+            future = self.thread_pool.submit(self._run_func, func, field)
+
+            res = "{%s}" % field
```

Comment:
please use f-strings

Reply:
the alternative f-string will be `f"{{{field}}}"`

since it needs to interpolate. having three curly braces would be difficult to understand what is happening. I will rewrite with string concat. OK?

Reply:
String concat is fine too

Reply:
done